File: wiki_data.py
import json

from model.aggressive_stats import AggressiveStats
from model.combat_stats import CombatStats
from model.defensive_stats import DefensiveStats
from model.locations import Location
from model.npc_stats import NpcStats
from model.weapon_category import WeaponCategory
from model.weapon_stats import WeaponStats


class WikiData:
    items_json = json.load(open("./wiki_data/items-dps-calc.json"))
    npcs_json = json.load(open("./wiki_data/npcs-dps-calc.json"))
    min_defence = json.load(open("./wiki_data/min_defence.json"))
    locations = json.load(open("./wiki_data/locations.json"))

    @staticmethod
    def get_item(item_id: int) -> WeaponStats:
        weapon = WikiData.items_json[str(item_id)]
        weapon_category = None if weapon.get("weaponCategory", 0) == 0 else WeaponCategory[weapon.get("weaponCategory", 0)]
        return WeaponStats(
            name=weapon.get("name", 0),
            id=weapon.get("id", item_id),
            attack_speed=weapon.get("aspeed", 0),
            stab=weapon.get("astab", 0),
            slash=weapon.get("aslash", 0),
            crush=weapon.get("acrush", 0),
            magic=weapon.get("amagic", 0),
            ranged=weapon.get("arange", 0),
            melee_strength=weapon.get("str", 0),
            ranged_strength=weapon.get("rstr", 0),
            magic_strength=weapon.get("mdmg", 0),
            weapon_category=weapon_category,
        )

    @staticmethod
    def get_npc(npc_id: int) -> NpcStats:
        npc = WikiData.npcs_json[str(npc_id)]
        npc_name = npc["name"]
        min_defence = WikiData.min_defence.get(npc_name, 0)
        location = WikiData.locations.get(npc_name, Location.NONE)

        return NpcStats(
            name=npc_name,
            combat_stats=CombatStats(
                hitpoints=npc.get("hitpoints", 0),
                attack=npc.get("att", 0),
                strength=npc.get("str", 0),
                defence=npc.get("def", 0),
                magic=npc.get("mage", 0),
                ranged=npc.get("range", 0),
            ),
            aggressive_stats=AggressiveStats(
                attack=npc.get("attbns", 0),
                magic=npc.get("amagic", 0),
                ranged=npc.get("arange", 0),
                melee_strength=npc.get("strbns", 0),
                ranged_strength=npc.get("rngbns", 0),
                magic_strength=npc.get("mbns", 0),
            ),
            defensive_stats=DefensiveStats(
                stab=npc.get("dstab", 0),
                slash=npc.get("dslash", 0),
                crush=npc.get("dcrush", 0),
                magic=npc.get("dmagic", 0),
                ranged=npc.get("drange", 0),
            ),
            min_defence=min_defence,
            location=location
        )

File: backend/osrs-dmg-sim/wikidata/wikidata.go
package wikidata

import (
	"embed"
	"encoding/json"
	"fmt"
)

type DataTypeProvider int

const (
	ItemProvider     DataTypeProvider = 1
	NpcProvider      DataTypeProvider = 2
	SpellProvider    DataTypeProvider = 3
	SpecProvider     DataTypeProvider = 4
	BisGraphProvider DataTypeProvider = 5
	IdAliasProvider  DataTypeProvider = 6
)

type ItemData struct {
	Name           string  `json:"name"`
	AttackSpeed    int     `json:"aspeed"`
	AStab          int     `json:"astab"`
	ASlash         int     `json:"aslash"`
	ACrush         int     `json:"acrush"`
	AMagic         int     `json:"amagic"`
	ARanged        int     `json:"arange"`
	DStab          int     `json:"dstab"`
	DSlash         int     `json:"dslash"`
	DCrush         int     `json:"dcrush"`
	DMagic         int     `json:"dmagic"`
	DRanged        int     `json:"drange"`
	MeleeStrength  int     `json:"str"`
	RangedStrength int     `json:"rstr"`
	MagicStrength  float32 `json:"mdmg"`
	Prayer         int     `json:"prayer"`
	Is2h           bool    `json:"is2h"`
	WeaponCategory string  `json:"weaponCategory"`
	Slot           int     `json:"slot"`
	Equipable      bool    `json:"equipable"`
}

type NpcData struct {
	Name      string `json:"name"`
	Hitpoints int    `json:"hitpoints"`
	Attack    int    `json:"att"`
	Strength  int    `json:"str"`
	Defence   int    `json:"def"`
	Magic     int    `json:"mage"`
	Ranged    int    `json:"range"`

	AAttack int `json:"attbns"`
	AMagic  int `json:"amagic"`
	ARange  int `json:"arange"`

	MeleeStrength  int `json:"strbns"`
	MagicStrength  int `json:"rngbns"`
	RangedStrength int `json:"mbns"`

	DStab                    int    `json:"dstab"`
	DSlash                   int    `json:"dslash"`
	DSCrush                  int    `json:"dcrush"`
	DMagic                   int    `json:"dmagic"`
	DRange                   int    `json:"drange"`
	DLight                   int    `json:"dlight"`
	DStandard                int    `json:"dstandard"`
	DHeavy                   int    `json:"dheavy"`
	ElementalWeaknessType    string `json:"elementalweaknesstype"`
	ElementalWeaknessPercent int    `json:"elementalweaknesspercent"`

	Size            int  `json:"size"`
	IsKalphite      bool `json:"isKalphite"`
	IsDemon         bool `json:"isDemon"`
	IsDragon        bool `json:"isDragon"`
	IsUndead        bool `json:"isUndead"`
	IsLeafy         bool `json:"isLeafy"`
	IsXerician      bool `json:"isXerician"`
	IsShade         bool `json:"isShade"`
	IsTobEntryMode  bool `json:"isTobEntryMode"`
	IsTobNormalMode bool `json:"isTobNormalMode"`
	IsTobHardMode   bool `json:"isTobHardMode"`
	Respawn         int  `json:"respawn"`
}

type SpellData struct {
	Name      string `json:"name"`
	Image     string `json:"image"`
	MaxHit    int    `json:"max_hit"`
	SpellBook string `json:"spellbook"`
	Element   string `json:"element"`
}

type BisSlotGraph struct {
	Nodes map[string]BisItem `json:"nodes"`
	Roots []string           `json:"roots"`
}

type BisItem struct {
	Ids  []string `json:"ids"`
	Next []string `json:"next"`
}

type BisGraphs map[string]map[string]BisSlotGraph

//go:embed json-data/*
var jsonDataEmbed embed.FS

func GetWikiData(p DataTypeProvider) any {
	var data any
	var err error

	switch p {
	case ItemProvider:
		data, err = getJsonData[map[int]ItemData]("json-data/items-dmg-sim.json")
	case NpcProvider:
		data, err = getJsonData[map[string]NpcData]("json-data/npcs-dmg-sim.json")
	case SpellProvider:
		data, err = getJsonData[[]SpellData]("json-data/spells.json")
	case SpecProvider:
		data, err = getJsonData[map[string]int]("json-data/special_attack.json")
	case BisGraphProvider:
		data, err = getJsonData[BisGraphs]("json-data/bis_graph.json")
	case IdAliasProvider:
		data, err = getJsonData[map[int]int]("json-data/id_aliases.json")
	}

	if err != nil {
		fmt.Println(err)
		return nil
	}
	return data
}

func getJsonData[T any](fileName string) (T, error) {
	var data T
	byteValue, err := jsonDataEmbed.ReadFile(fileName)

	if err != nil {
		return data, err
	}

	if err := json.Unmarshal(byteValue, &data); err != nil {
		return data, err
	}
	return data, err
}
